### PR TITLE
Kernel<-Fusion Step 3 - Debug print for expr_eval and type promotion fix

### PR DESCRIPTION
**Pull request by csarofeen**

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp [hunk 1]
```diff
@@ -79,6 +79,10 @@ void ExpressionEvaluator::print() const {
   for (const auto& kv : known_values_) {
     std::cout << kv.first->toString() << " = " << kv.second << "\n";
   }
+  std::cout << "--------------------\n";
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp]: Could we have a line above this like `std::cout << "\Pre-computed Values\n";` ?

File: torch/csrc/jit/codegen/cuda/type.cpp [hunk 2]
```diff
@@ -87,6 +87,9 @@ ValType promote_type(const ValType& t1, const ValType& t2) {
       (t1 == ValType::Scalar || t1 == ValType::NamedScalar)) {
     return ValType::Scalar;
   }
+  if (t1 == ValType::NamedScalar && t2 == ValType::NamedScalar) {
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/type.cpp:90]: Looks like an important fix 

I'm guessing we probably don't want it in this PR though.

**Review by shmsong — APPROVED** [review]
LGTM. Some minor comments. 

This is very useful debug feature. Thanks for building it 👍

File: torch/csrc/jit/codegen/cuda/type.cpp [hunk 3]
```diff
@@ -87,6 +87,9 @@ ValType promote_type(const ValType& t1, const ValType& t2) {
...
+    return ValType::Scalar;
+  }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/type.cpp:92]: Looks like `promote_type` should just return `ValType::Scalar` whenever none of the inputs is `TensorView`.

File: torch/csrc/jit/codegen/cuda/type.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/type.cpp:90]: The change makes sense to me. I haven't seen an exception here before, so looks like there has just been no such case so far.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/type.cpp:90]: The exception showed up in the following PR to this one. I'm too lazy to split it out into a separate PR line so leaving it in here.

File: torch/csrc/jit/codegen/cuda/type.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/type.cpp:92]: Likely, but want to keep this narrow and assert on any unexpected val types sent as someone could send something like TensorIndex which wouldn't make any sense here.

**Pull request merged**